include/treespec.h

```diff
@@ -201,17 +204,34 @@ class PyTreeSpec {
                 INTERNAL_ERROR();
         }
 
-        h = H::combine(
+        return H::combine(
             std::move(h), n.kind, n.arity, n.custom, n.num_leaves, n.num_nodes, data_hash);
-        return h;
+    }
+
+    template <typename H>
+    friend H AbslHashValueImpl(H h, const PyTreeSpec &t) {
+        return H::combine(std::move(h), t.m_traversal, t.m_none_is_leaf, t.m_namespace);
     }
 
     template <typename H>
     friend H AbslHashValue(H h, const PyTreeSpec &t) {
```

Comment:
Why friend here

Reply:
This is required by the abseil-cpp hash library. Otherwise, it will raise `PoisonedHash` error during the compilation time.

Reply:
I checked [official docs](https://abseil.io/docs/cpp/guides/hash). Seems to be an implementation of a function rather than a class method.